api/src/main/java/org/pmiops/workbench/impersonation/ImpersonatedFirecloudServiceImpl.java

```diff
@@ -0,0 +1,49 @@
+package org.pmiops.workbench.impersonation;
+
+import static org.pmiops.workbench.firecloud.FireCloudServiceImpl.TERMS_OF_SERVICE_BODY;
+
+import java.io.IOException;
+import javax.annotation.Nonnull;
+import org.pmiops.workbench.db.model.DbUser;
+import org.pmiops.workbench.firecloud.FirecloudApiClientFactory;
+import org.pmiops.workbench.firecloud.FirecloudRetryHandler;
+import org.pmiops.workbench.firecloud.api.TermsOfServiceApi;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+/**
+ * An impersonation-enabled version of {@link org.pmiops.workbench.firecloud.FireCloudServiceImpl}
+ *
+ * <p>REMINDER: With great power comes great responsibility. Impersonation should not be used in
+ * production, except where absolutely necessary.
+ */
+@Service
+public class ImpersonatedFirecloudServiceImpl implements ImpersonatedFirecloudService {
+
+  private final FirecloudApiClientFactory firecloudApiClientFactory;
+  private final FirecloudRetryHandler retryHandler;
+
+  @Autowired
+  public ImpersonatedFirecloudServiceImpl(
+      FirecloudApiClientFactory firecloudApiClientFactory, FirecloudRetryHandler retryHandler) {
+    this.firecloudApiClientFactory = firecloudApiClientFactory;
+    this.retryHandler = retryHandler;
+  }
+
+  @Override
+  public void acceptTermsOfService(@Nonnull DbUser dbUser) throws IOException {
+    TermsOfServiceApi termsOfServiceApi = getTosApi(dbUser);
```

Comment:
can we log  the logged in user's information so that we can know who is doing the impersonation for our records?

Reply:
There's no logged-in user here - it's the App SA.

Do you mean the user that is being impersonated?  I'm logging that in OfflineTestUsersController - would you prefer to see that here?

Reply:
I assumed that its being used apart from cron job as well.  That is fine then